File: XPath/org/eclipse/wst/xml/xpath2/processor/util/StaticContextBuilder.cs
using System.Collections;
using System.Collections.Generic;
using java.net;
using java.xml;
using javax.xml.@namespace;

/// <summary>
///*****************************************************************************
/// Copyright (c) 2011, Jesper Steen Moller, and others
/// All rights reserved. This program and the accompanying materials
/// are made available under the terms of the Eclipse Public License 2.0
/// which accompanies this distribution, and is available at
/// https://www.eclipse.org/legal/epl-2.0/
/// 
/// SPDX-License-Identifier: EPL-2.0
/// 
/// Contributors:
///     Jesper Steen Moller - initial API and implementation
///     Jesper Steen Moller  - bug 340933 - Migrate to new XPath2 API
///     Jesper Steen Moller - bug 343804 - Updated API information
/// ******************************************************************************
/// </summary>

namespace org.eclipse.wst.xml.xpath2.processor.util
{
	using URI = java.net.URI;
	using XMLConstants = java.xml.XMLConstants;
	using NamespaceContext = javax.xml.@namespace.NamespaceContext;
	using QName = javax.xml.@namespace.QName;
	using CollationProvider = org.eclipse.wst.xml.xpath2.api.CollationProvider;
	using Function = org.eclipse.wst.xml.xpath2.api.Function;
	using FunctionLibrary = org.eclipse.wst.xml.xpath2.api.FunctionLibrary;
	using StaticContext = org.eclipse.wst.xml.xpath2.api.StaticContext;
	using StaticVariableResolver = org.eclipse.wst.xml.xpath2.api.StaticVariableResolver;
	using ItemType = org.eclipse.wst.xml.xpath2.api.typesystem.ItemType;
	using TypeDefinition = org.eclipse.wst.xml.xpath2.api.typesystem.TypeDefinition;
	using TypeModel = org.eclipse.wst.xml.xpath2.api.typesystem.TypeModel;
	using FnFunctionLibrary = org.eclipse.wst.xml.xpath2.processor.function.FnFunctionLibrary;
	using XSCtrLibrary = org.eclipse.wst.xml.xpath2.processor.function.XSCtrLibrary;
	using BuiltinTypeLibrary = org.eclipse.wst.xml.xpath2.processor.@internal.types.builtin.BuiltinTypeLibrary;
	using Node = org.w3c.dom.Node;

	/// <summary>
	/// @since 2.0
	/// </summary>
	public class StaticContextBuilder : StaticContext
	{
		private bool InstanceFieldsInitialized = false;

		public StaticContextBuilder()
		{
			if (!InstanceFieldsInitialized)
			{
				InitializeInstanceFields();
				InstanceFieldsInitialized = true;
			}
            _collationProvider = new CollationProviderAnonymousInnerClass(this);
		}

		private void InitializeInstanceFields()
		{
			_functionLibraries[XPATH_FUNCTIONS_NS] = new FnFunctionLibrary();
			_functionLibraries[XSCtrLibrary.XML_SCHEMA_NS] = new XSCtrLibrary();
		}


		public const string XPATH_FUNCTIONS_NS = "http://www.w3.org/2005/xpath-functions";

		private bool _xpath1_compatible = false;
		private string _default_namespace = "";
		private string _default_function_namespace = XPATH_FUNCTIONS_NS;
		private TypeDefinition _initialContextType = null;
		private string _defaultCollation = org.eclipse.wst.xml.xpath2.api.CollationProvider_Fields.CODEPOINT_COLLATION;

		// key: String prefix, contents: String namespace
		private Dictionary<string, string> _namespaces = new Dictionary<string, string>();
		private IDictionary<string, FunctionLibrary> _functionLibraries = new Dictionary<string, FunctionLibrary>();

		private URI _base_uri;
		private IDictionary _variableTypes = new Hashtable();
		private IDictionary _variableCardinality = new Hashtable();
		private IDictionary _collectionTypes = new Hashtable();

		private HashSet<QName> _hiddenFunctions = new HashSet<QName>();

		private TypeModel _typeModel;

		public virtual bool XPath1Compatible
		{
			get
			{
				return _xpath1_compatible;
			}
		}

		public virtual NamespaceContext NamespaceContext
		{
			get
			{
				return new NamespaceContextAnonymousInnerClass(this);
			}
		}

		private class NamespaceContextAnonymousInnerClass : NamespaceContext
		{
			private readonly StaticContextBuilder outerInstance;

			public NamespaceContextAnonymousInnerClass(StaticContextBuilder outerInstance)
			{
				this.outerInstance = outerInstance;
			}


			public virtual IEnumerator<string> getPrefixes(string ns)
			{
				var prefixes = new List<string>();
				for (var it = outerInstance._namespaces.GetEnumerator(); it.MoveNext();)
				{
					var entry = it.Current;

					if (entry.Value.Equals(ns))
					{
						prefixes.Add(entry.Key);
					}
				}
				return prefixes.GetEnumerator();
			}

			public virtual string getPrefix(string ns)
			{
				for (var it = outerInstance._namespaces.GetEnumerator(); it.MoveNext();)
				{
					var entry = it.Current;
					if (entry.Value.Equals(ns))
					{
						return entry.Key;
					}
				}
				return null;
			}

			public virtual string getNamespaceURI(string prefix)
			{
				string ns = outerInstance._namespaces[prefix];
				if (string.ReferenceEquals(ns, null))
				{
					ns = XMLConstants.NULL_NS_URI;
				}
				return ns;
			}
		}

		public virtual string DefaultNamespace
		{
			get
			{
				return _default_namespace;
			}
		}

		public virtual string DefaultFunctionNamespace
		{
			get
			{
				return _default_function_namespace;
			}
		}

		public virtual TypeModel TypeModel
		{
			get
			{
				if (_typeModel != null)
				{
					return _typeModel;
				}
    
				return new TypeModelAnonymousInnerClass(this);
			}
		}

		private class TypeModelAnonymousInnerClass : TypeModel
		{
			private readonly StaticContextBuilder outerInstance;

			public TypeModelAnonymousInnerClass(StaticContextBuilder outerInstance)
			{
				this.outerInstance = outerInstance;
			}


			public virtual TypeDefinition getType(Node node)
			{
				return null;
			}

			public virtual TypeDefinition lookupType(string @namespace, string typeName)
			{
				return null;
			}

			public virtual TypeDefinition lookupElementDeclaration(string @namespace, string elementName)
			{
				return null;
			}

			public virtual TypeDefinition lookupAttributeDeclaration(string @namespace, string attributeName)
			{
				return null;
			}
		}

		public virtual Function resolveFunction(QName name, int arity)
		{
			if (_hiddenFunctions.Contains(name))
			{
				return null;
			}
			FunctionLibrary flib = _functionLibraries[name.NamespaceURI];
			if (flib != null)
			{
				return flib.resolveFunction(name.LocalPart, arity);
			}
			return null;
		}

		public virtual URI BaseUri
		{
			get
			{
				return _base_uri;
			}
		}

		public virtual IDictionary<string, FunctionLibrary> FunctionLibraries
		{
			get
			{
				return _functionLibraries;
			}
		}

		public virtual TypeDefinition getCollectionType(string collectionName)
		{
			return (TypeDefinition) _collectionTypes[collectionName];
		}

		public virtual TypeDefinition InitialContextType
		{
			get
			{
				return _initialContextType;
			}
		}

		public virtual StaticContextBuilder withNamespace(string prefix, string uri)
		{
			_namespaces[prefix] = uri;
			return this;
		}

		public virtual StaticContextBuilder withDefaultNamespace(string uri)
		{
			_default_namespace = uri;
			return this;
		}

		public virtual StaticContextBuilder withXPath1Compatibility(bool compatible)
		{
			_xpath1_compatible = compatible;
			return this;
		}

		public virtual StaticContextBuilder withTypeModel(TypeModel tm)
		{
			_typeModel = tm;
			return this;
		}

		public virtual StaticContextBuilder withoutFunction(QName functionToSuppress)
		{
			_hiddenFunctions.Add(functionToSuppress);
			return this;
		}
		public virtual StaticContextBuilder withoutFunction(params QName[] functionsToSuppress)
		{
			foreach (QName name in functionsToSuppress)
			{
				_hiddenFunctions.Add(name);
			}
			return this;
		}

		public virtual TypeDefinition DefaultCollectionType
		{
			get
			{
				return BuiltinTypeLibrary.XS_UNTYPED;
			}
		}

		public virtual StaticVariableResolver InScopeVariables
		{
			get
			{
				return new StaticVariableResolverAnonymousInnerClass(this);
			}
		}

		private class StaticVariableResolverAnonymousInnerClass : StaticVariableResolver
		{
			private readonly StaticContextBuilder outerInstance;

			public StaticVariableResolverAnonymousInnerClass(StaticContextBuilder outerInstance)
			{
				this.outerInstance = outerInstance;
			}


			public virtual bool isVariablePresent(QName name)
			{
				return outerInstance._variableTypes.Contains(name);
			}

			public virtual ItemType getVariableType(QName name)
			{
				return (ItemType) outerInstance._variableTypes[name];
			}
		}

		// We are explicitly NOT using generics here, in anticipation of JDK1.4 compatibility
		private static IComparer CODEPOINT_COMPARATOR = new ComparatorAnonymousInnerClass();

		private class ComparatorAnonymousInnerClass : IComparer
		{
			public ComparatorAnonymousInnerClass()
			{
			}


			public virtual int compare(object o1, object o2)
			{
				return ((string)o1).CompareTo((string)o2);
			}

            public int Compare(object o1, object o2)
            {
                return ((string)o1).CompareTo((string)o2);
            }
		}

        private CollationProvider _collationProvider = null;

		private class CollationProviderAnonymousInnerClass : CollationProvider
		{
			public CollationProviderAnonymousInnerClass(StaticContextBuilder builder)
            {
                DefaultCollation = builder._defaultCollation;
            }


			public virtual string DefaultCollation
			{
				get; private set;
			}

			public virtual IComparer<string> getCollation(string uri)
			{
				//if (org.eclipse.wst.xml.xpath2.api.CollationProvider_Fields.CODEPOINT_COLLATION.Equals(uri))
				//{
				//	return CODEPOINT_COMPARATOR;
				//}
				return null;
			}
		}

		public virtual CollationProvider CollationProvider
		{
			get
			{
				return _collationProvider;
			}
		}

		public virtual StaticContextBuilder withCollationProvider(CollationProvider cp)
		{
			_collationProvider = cp;
			return this;
		}

		public virtual StaticContextBuilder withVariable(QName qName, ItemType type)
		{
			_variableTypes[qName] = type;
			return this;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in .NET:
//ORIGINAL LINE: public StaticContextBuilder withBaseUri(String string) throws java.net.URISyntaxException
		public virtual StaticContextBuilder withBaseUri(string @string)
		{
			_base_uri = new URI(@string);
			return this;
		}

		public virtual StaticContextBuilder withFunctionLibrary(string @namespace, FunctionLibrary fl)
		{
			_functionLibraries[@namespace] = fl;
			return this;
		}

		public virtual StaticContextBuilder withDefaultCollation(string uri)
		{
			this._defaultCollation = uri;
			return this;
		}

		public virtual ItemType getDocumentType(URI documentUri)
		{
			return null;
		}
	}

}

File: AntlrTreeEditing/src-java/org/eclipse/wst/xml/xpath2/processor/util/StaticContextBuilder.java
/*******************************************************************************
 * Copyright (c) 2011, Jesper Steen Moller, and others
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Jesper Steen Moller - initial API and implementation
 *     Jesper Steen Moller  - bug 340933 - Migrate to new XPath2 API
 *     Jesper Steen Moller - bug 343804 - Updated API information
 *******************************************************************************/

package org.eclipse.wst.xml.xpath2.processor.util;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.xml.XMLConstants;
import javax.xml.namespace.NamespaceContext;
import javax.xml.namespace.QName;

import org.eclipse.wst.xml.xpath2.api.CollationProvider;
import org.eclipse.wst.xml.xpath2.api.Function;
import org.eclipse.wst.xml.xpath2.api.FunctionLibrary;
import org.eclipse.wst.xml.xpath2.api.StaticContext;
import org.eclipse.wst.xml.xpath2.api.StaticVariableResolver;
import org.eclipse.wst.xml.xpath2.api.typesystem.ItemType;
import org.eclipse.wst.xml.xpath2.api.typesystem.TypeDefinition;
import org.eclipse.wst.xml.xpath2.api.typesystem.TypeModel;
import org.eclipse.wst.xml.xpath2.processor.function.FnFunctionLibrary;
import org.eclipse.wst.xml.xpath2.processor.function.XSCtrLibrary;
import org.eclipse.wst.xml.xpath2.processor.internal.types.builtin.BuiltinTypeLibrary;
import org.w3c.dom.Node;

/**
 * @since 2.0
 */
public class StaticContextBuilder implements StaticContext {

	public static final String XPATH_FUNCTIONS_NS = "http://www.w3.org/2005/xpath-functions";

	private boolean _xpath1_compatible = false;
	private String _default_namespace = "";
	private String _default_function_namespace = XPATH_FUNCTIONS_NS;
	private TypeDefinition _initialContextType = null; 
	private String _defaultCollation = CollationProvider.CODEPOINT_COLLATION;

	// key: String prefix, contents: String namespace
	private Map/*<String, String>*/ _namespaces = new HashMap/*<String, String>*/();
	private Map<String, FunctionLibrary> _functionLibraries = new HashMap<String, FunctionLibrary>();
	{
		_functionLibraries.put(XPATH_FUNCTIONS_NS, new FnFunctionLibrary());
		_functionLibraries.put(XSCtrLibrary.XML_SCHEMA_NS, new XSCtrLibrary());
	}

	private URI _base_uri;
	private Map/*<String, TypeDefinition>*/ _variableTypes = new HashMap/*<String, TypeDefinition>*/();
	private Map/*<String, Short>*/ _variableCardinality = new HashMap/*<String, Short>*/();
	private Map/*<String, TypeDefinition>*/ _collectionTypes = new HashMap/*<String, TypeDefinition>*/();

	private Set/*<QName>*/ _hiddenFunctions = new HashSet/*<QName>*/();

	private TypeModel _typeModel;
	
	public boolean isXPath1Compatible() {
		return _xpath1_compatible;
	}

	public NamespaceContext getNamespaceContext() {
		return new NamespaceContext() {
			
			public Iterator getPrefixes(String ns) {
				List/*<String>*/ prefixes = new LinkedList/*<String>*/();
				for (Iterator it = _namespaces.entrySet().iterator(); it.hasNext(); ) {
					Map.Entry entry = (Map.Entry)it.next();
					
					if (entry.getValue().equals(ns)) prefixes.add(entry.getKey());
				}
				return prefixes.iterator();
			}
			
			public String getPrefix(String ns) {
				for (Iterator it = _namespaces.entrySet().iterator(); it.hasNext(); ) {
					Map.Entry entry = (Map.Entry)it.next();					
					if (entry.getValue().equals(ns)) {
						return (String)entry.getKey();
					}
				}
				return null;
			}
			
			public String getNamespaceURI(String prefix) {
				String ns = (String)_namespaces.get(prefix);
				if (ns == null) ns = XMLConstants.NULL_NS_URI;
				return ns;
			}
		};
	}

	public String getDefaultNamespace() {
		return _default_namespace;
	}

	public String getDefaultFunctionNamespace() {
		return _default_function_namespace;
	}

	public TypeModel getTypeModel() {
		if (_typeModel != null) return _typeModel;
		
		return new TypeModel() {

			public TypeDefinition getType(Node node) {
				return null;
			}

			public TypeDefinition lookupType(String namespace, String typeName) {
				return null;
			}

			public TypeDefinition lookupElementDeclaration(String namespace, String elementName) {
				return null;
			}

			public TypeDefinition lookupAttributeDeclaration(String namespace, String attributeName) {
				return null;
			}};
	}

	public Function resolveFunction(QName name, int arity) {
		if (_hiddenFunctions.contains(name)) return null;
		FunctionLibrary flib = _functionLibraries.get(name.getNamespaceURI());
		if (flib != null) {
			return flib.resolveFunction(name.getLocalPart(), arity);
		}
		return null;
	}

	public URI getBaseUri() {
		return _base_uri;
	}

	public Map<String, FunctionLibrary> getFunctionLibraries() {
		return _functionLibraries;
	}

	public TypeDefinition getCollectionType(String collectionName) {
		return (TypeDefinition) _collectionTypes.get(collectionName);
	}

	public TypeDefinition getInitialContextType() {
		return _initialContextType;
	}

	public StaticContextBuilder withNamespace(String prefix, String uri) {
		_namespaces.put(prefix, uri);
		return this;
	}

	public StaticContextBuilder withDefaultNamespace(String uri) {
		_default_namespace = uri;
		return this;
	}

	public StaticContextBuilder withXPath1Compatibility(boolean compatible) {
		_xpath1_compatible = compatible;
		return this;
	}

	public StaticContextBuilder withTypeModel(TypeModel tm) {
		_typeModel = tm;
		return this;
	}

	public StaticContextBuilder withoutFunction(QName functionToSuppress) {
		_hiddenFunctions.add(functionToSuppress);
		return this;
	}
	public StaticContextBuilder withoutFunction(QName ... functionsToSuppress) {
		for (QName name : functionsToSuppress)
			_hiddenFunctions.add(name);
		return this;
	}
	
	public TypeDefinition getDefaultCollectionType() {
		return BuiltinTypeLibrary.XS_UNTYPED;
	}

	public StaticVariableResolver getInScopeVariables() {
		return new StaticVariableResolver() {

			public boolean isVariablePresent(QName name) {
				return _variableTypes.containsKey(name);
			}

			public ItemType getVariableType(QName name) {
				return (ItemType) _variableTypes.get(name);
			}
		};
	}

	// We are explicitly NOT using generics here, in anticipation of JDK1.4 compatibility
	private static Comparator CODEPOINT_COMPARATOR = new Comparator() {
		
		public int compare(Object o1, Object o2) {
			return ((String)o1).compareTo((String)o2);
		}
	};
	
	private CollationProvider _collationProvider = new CollationProvider() {
		
		public String getDefaultCollation() {
			return _defaultCollation;
		}
		
		public Comparator getCollation(String uri) {
			if (CollationProvider.CODEPOINT_COLLATION.equals(uri)) return CODEPOINT_COMPARATOR;
			return null;
		}
	};

	public CollationProvider getCollationProvider() {
		return _collationProvider;
	}

	public StaticContextBuilder withCollationProvider(CollationProvider cp) {
		_collationProvider = cp;
		return this;
	}

	public StaticContextBuilder withVariable(javax.xml.namespace.QName qName,
			ItemType type) {
		_variableTypes.put(qName, type);
		return this;
	}

	public StaticContextBuilder withBaseUri(String string) throws URISyntaxException {
		_base_uri = new URI(string);		
		return this;
	}

	public StaticContextBuilder withFunctionLibrary(
			String namespace,
			FunctionLibrary fl) {
		_functionLibraries.put(namespace, fl);
		return this;
	}

	public StaticContextBuilder withDefaultCollation(String uri) {
		this._defaultCollation = uri;
		return this;
	}

	public ItemType getDocumentType(URI documentUri) {
		return null;
	}
}
